src/osd/PrimaryLogPG.cc

```diff
@@ -3251,24 +3262,116 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
-void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, object_locator_t oloc, hobject_t soid,
-                                     SnapContext snapc, bool get, RefCountCallback *cb, uint64_t offset)
+bool PrimaryLogPG::check_has_ref_snap(hobject_t soid, snapid_t snapid, chunk_info_t& ci)
+{
+  hobject_t clone_oid = soid;
+  clone_oid.snap = snapid;
+  bool has_reference = false;
+  ObjectContextRef cobc = get_object_context(clone_oid, false, NULL);
+  if (!cobc) {
+    dout(0) << __func__ << ": Can not find clone obc " << clone_oid << dendl;;
+    return false;
+  }
+  object_info_t& coi = cobc->obs.oi;
+  for (auto c : coi.manifest.chunk_map) {
+    if (c.second == ci) {
+      has_reference = true;
+      break;
+    }
+  }
+  dout(15) << __func__ << " target offset: " << ci.offset  << " oid: " << ci.oid 
+	  << " has references " << has_reference << dendl;
+  return has_reference;
+}
+
+void PrimaryLogPG::dec_refcount_non_intersection(ObjectContextRef obc, const object_info_t& oi, 
+						  set<uint64_t> intersection_set)
+{
+  for (auto c : oi.manifest.chunk_map) {
+    auto iter = intersection_set.find(c.first);
+    if (intersection_set.end() == iter) {
+      dout(10) << __func__ << " need dereference. " << " offset: " <<  c.first 
+	      << " length: " << c.second.length << " oid: " << c.second.oid << dendl;
+      object_locator_t target_oloc(c.second.oid);
+      refcount_manifest(obc, obc->obs.oi.soid, target_oloc, c.second.oid, 
+			SnapContext(), false, NULL);
+    }
+  }
+}
+
+void PrimaryLogPG::dec_all_refcount_manifest(object_info_t& oi, OpContext* ctx)
+{
+  SnapSetContext* ssc = ctx->obc->ssc;
+  ceph_assert(oi.has_manifest());
```

Comment:
This method only gets called on the head object, right?  Can we change the name to reflect that and add an assert?

Reply:
Ok.